More hooks

tiosgz opened the pull request:
* automatically expand window panel tree to active tab &/ select the active tab (triggered on tab change)
* encode search string when keyword-searching from QC ([VB-59967](https://forum.vivaldi.net/topic/42066/search-engine-in-quick-commands-with-nickname-cannot-recognize))
* change tree view row height (as mentioned there, with some extremes it requires user CSS to look nice)
* resize tabs (only for horizontal tab bar)

tiosgz commented on vivaldi/hooks/auto-expand-window-tree.js:
```diff
@@ -0,0 +1,21 @@
+                // Expand tree to active tab
+                this.refs.treeList.expandToId(e.id)
+                // Change selection to active tab
+                // this.refs.treeList.selectById(e.id)
```
I guess you want both to be toggle-able from settings?

tiosgz commented:
So I think now I’ve gotten to the situation when a newbie writes
```ruby
puts 'Hello!'
puts 'Hello!'
puts 'Hello!'
```
and when you ask them if they know what’s wrong with it, they answer they don’t.

That said, I think I’m done, but there’s surely a lot to improve.

And, not to forget, `encoded-qc-search.js` may (or may not) get useless with next snappy, so I’d wait for it if all other is done earlier.

tiosgz commented:
I was wondering how to make `tab-size.js` allow really wide tabs (e.g. to fill all the space) w/o cluttering it even more or decreasing the sliders’ precision (it’s now already quite hard to choose the correct size with mouse). Do you have any idea?

justdanpo commented:
You can  tread maximum value as "auto".
Ok, what about latest two of my comments? Have you seen them?

tiosgz commented:
> You can tread maximum value as "auto".

Of course. Thanks!
> Ok, what about latest two of my comments? Have you seen them?

Sorry, I don’t think so. What? Where? Who? I really don’t know.

justdanpo commented:
Oh, I thought you haven't reacted on my comments while that was my fault. I'm sorry. The same thing with the previous pr.

tiosgz commented on vivaldi/hooks/encoded-qc-search.js:
```diff
@@ -0,0 +1,37 @@
+        constructor(...e) {
+            super(...e)
+
+            this.getKeywordSearch = query => {
```
This wouldn’t work so well, coz you don’t want the user to see `Search “1%251”…`, but `Search “1+1”…`

I changed it to use the original function though

justdanpo replied:
Oh, I haven't checked how it's displayed.

tiosgz commented on vivaldi/hooks/treeview-spacing.js:
```diff
@@ -0,0 +1,87 @@
+vivaldi.jdhooks.hookClass('common_VivaldiTreeList', cls => {
+    const newCls = vivaldi.jdhooks.insertWatcher(class extends cls {
+        render() {
+            this.props.rowHeight = this.state.jdVivaldiSettings.VIVALDI_TREE_ROW_HEIGHT
```
So what you want me to do is to hook all the classes that use it (should be 7 in total), right? I think adding settings per tree type would be easier and better for users (coz one may want different size in bookmarks & in notes etc.).

I’m not sure if (& how) your suggestions would work. How do you mean the function? That would have to be called by me, or not?

justdanpo replied:
I proposed 2 ways:
- change only that items which height is equal to default
- use multiplication by some floating value to change height of every item

May be there are other solutions exist.

Talking about a function I mean `props.rowHeight` may be a function returning height by item's index, so items of the same tree may have different heights. So you need to ignore values if their `typeof` is "function" or make overridden value a function calling original one and modifying its return value. 
I haven't seen this feature is used anywhere but may be it will appear when mail or some other new feature is published.

tiosgz replied:
Sorry for the delay, I’m quite unfocused recently. It seems that I can only change the spacing from the `TreeList` itself, so I have to find a way to differentiate the trees—it seems *most* of them have either unique `props.mimeType` or `props.settings`, so I think this is the way to go. If you have any suggestions, please let me know.

(Just a progress report)

tiosgz replied:
Can it be like this? https://github.com/justdanpo/VivaldiHooks/pull/37/commits/40778aefd6de6f6f8555ae07a4f82ebe6cc5871a

tiosgz commented:
Just wondering… is that those invisible comments again or have you just not replied yet?

justdanpo commented:
> Just wondering… is that those invisible comments again or have you just not replied yet?

No, sorry, just a tough week at work. I'm exhausted. Waiting for the weekend.

tiosgz commented:
> No, sorry, just a tough week at work. I'm exhausted. Waiting for the weekend.

OK, thanks for the info

justdanpo commented on vivaldi/hooks/treeview-spacing.js:
```diff
@@ -0,0 +1,83 @@
+        render() {
+            const rowHeight = this.state.jdVivaldiSettings.VIVALDI_TREE_ROW_HEIGHT
+
+            return React.createElement('div', { className: 'setting-group' },
```
I have github reviews =(

tiosgz commented on vivaldi/hooks/window-panel-tab-switched-actions.js:
```diff
@@ -0,0 +1,83 @@
+                    React.createElement('span', null, label)))
+        }
+        render() {
+            return React.createElement('div', { className: 'setting-group' },
```
Thanks! This is how it looks like when I copy it from the most recent hook instead of the most relevant one :-(

tiosgz commented:
May I remind you of this PR... ?

justdanpo commented:
I'm so sorry, I still cannot check your PR because of total lack of energy =(

tiosgz commented:
Just rebased & instead of moving `encoded-qc-search.js` to archive, I deleted it (like you’ve deleted archive in https://github.com/justdanpo/VivaldiHooks/commit/d3a934e01e173b2359b2dc302f564807198a1128, but I didn’t feel like keeping the old commit in a PR)

justdanpo commented on vivaldi/hooks/tab-size.js:
```diff
@@ -0,0 +1,127 @@
+// Change the width of a tab
+
+vivaldi.jdhooks.hookModule("vivaldiSettings", (moduleInfo, exports) => {
+    let oldGetDefault = exports.getDefault
```
It's strange I've not used const in such overrides. Ok, I'll do it later.

justdanpo commented on vivaldi/hooks/treeview-spacing.js:
```diff
@@ -0,0 +1,184 @@
+`, 'treeview-spacing.js');
+
+// Odd name to avoid conflicts
+const jdHooksTvwSpcDetails = [
```
You may use classical js trick to avoid name conflicts:

```javascript
(function(){
//the whole file content except the initial comment describing the mod
})()
```

tiosgz replied:
Thanks, will do. I don’t know why I didn’t use it.

justdanpo commented on vivaldi/hooks/window-panel-tab-switched-actions.js:
```diff
@@ -0,0 +1,86 @@
+            let old_onPageStoreChanged = this._onPageStoreChanged
+            this._onPageStoreChanged = () => {
+                old_onPageStoreChanged()
+                const e = PageStore.a.getActivePage()
```
Remove `.a`
I've rewritten `require` so it uses `exports` member of `moduleSignatures` (see jdhooks.js) which potentially allows specifying which of these `.a`, `.b` etc to select by meaningful name ("default" is default one :) ). I'll update wiki later.

tiosgz commented:
Hmm, while testing the hooks I found out that `treeview-spacing` doesn’t work for the menu editor — & that there’s none of the currently used ways of identification available. I’ll try & look for another later today… (if you know of one, please tell me)
